### PR TITLE
core: Add ResourceTable

### LINKED ISSUE #3134: refactor: Resource table
Before working on #2180 we need to address situation with resource table. Resource table is used to store resource ids with their corresponding representations. 

https://github.com/denoland/deno/blob/b1685ce0101f9ca16b0eeb6db7ff16ac81301545/cli/resources.rs#L56-L104

There are two problems with this approach:
- all resource related logic lives in `cli/resources.rs`, instead logic for each kind of resource should live alongside ops code that utilizes it
- all representations are hard coded which effectively prohibits splitting ops into separate crates - CLI would still have to be aware what kind of resource each op can use

We want `ResourceTable` to expose public API allowing op crates to register any resources they see fit to utilize (eg. `process` crate would implement resources analogous to `Repr::Child`, `Repr:ChildStdin`, `Repr:ChildStdout` and `Repr:ChildStderr`). 

That means that `ResourceTable` should be a struct mapping resource ids to boxed trait objects. The trait could be as simple as:
```rust
trait Resource {
  pub fn close(&self);
}

struct ResourceTable(HashMap<Rid, Box<dyn Resource>>);
```
Trait objects from `ResourceTable` would then be upcasted into concrete type inside the op.

New resource table should be a `core` concept. It should be first implemented for `deno_core_http_bench`, then we should refactor `cli::resources::RESOURCE_TABLE`.

I'll be working on it.

CC @ry @piscisaureus

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #3134 

CC @ry

### CODE REVIEW

File: core/resources.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  pub fn close(&mut self, rid: ResourceId) -> Result<(), Error> {
+    let repr = self.map.remove(&rid).ok_or_else(bad_resource)?;
+    // Give resource a chance to cleanup (notify tasks, etc.)
+    repr.close();
+    Ok(())
+  }
+}
+
+/// Abstract type representing resource in Deno.
+pub trait Resource: Downcast + Any + Send {
```
Comment (bartlomieju) [inline core/resources.rs]: I tried implementing downcasting manually, but got such errors:
```
error[E0038]: the trait `resources::Resource` cannot be made into an object
  --> core/resources.rs:61:3
   |
61 |   pub fn add(&mut self, resource: Box<dyn Resource>) -> ResourceId {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `resources::Resource` cannot be made into an object
   |
   = note: method `downcast_ref` has generic type parameters
   = note: method `downcast_mut` has generic type parameters
```
That's why I used `Denocast` trait from `downcast-rs` crate.

File: core/examples/http_bench.rs [hunk 2]
```diff
@@ -184,115 +182,93 @@ fn main() {
...
-      }
-    })
-    .and_then(move |nwritten| {
-      debug!("write success {}", nwritten);
-      Ok(nwritten as i32)
-    }),
-  )
+  let fut = futures::future::poll_fn(move || {
+    let mut table = get_resource_table();
+    let stream = table.get_mut::<TcpStream>(rid)?;
```
Comment (bartlomieju) [inline core/examples/http_bench.rs]: Would be awesome if we could do something like:
```
let stream = table.get_mut::<dyn AsyncWrite>(rid)?;
stream.poll_write(&zero_copy_buf)
```

File: core/examples/http_bench.rs [hunk 3]
```diff
@@ -184,115 +182,93 @@ fn main() {
...
-  static ref NEXT_RID: AtomicUsize = AtomicUsize::new(3);
+  static ref RESOURCE_TABLE: Mutex<ResourceTable> =
+    Mutex::new(ResourceTable::default());
 }
 
-fn new_rid() -> i32 {
-  let rid = NEXT_RID.fetch_add(1, Ordering::SeqCst);
-  rid as i32
+fn get_resource_table<'a>() -> MutexGuard<'a, ResourceTable> {
+  RESOURCE_TABLE.lock().unwrap()
```
Comment (ry) [inline core/examples/http_bench.rs]: When we port this to CLI, I think the resource table should be owned by Worker.

Maybe s/get_resource_table/lock_resource_table/

File: core/resources.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+
+// Think of Resources as File Descriptors. They are integers that are allocated
+// by the privileged side of Deno to refer to various resources.  The simplest
+// example are standard file system files and stdio - but there will be other
+// resources added in the future that might not correspond to operating system
+// level File Descriptors. To avoid confusion we call them "resources" not "file
+// descriptors". This module implements a global resource table. Ops (AKA
+// handlers) look up resources by their integer id here.
```
Comment (ry) [inline core/resources.rs]: Please replace the comment with this edit:

> Think of Resources as File Descriptors. They are integers that are allocated by the privileged side of Deno to refer to various rust objects that need to be referenced between multiple ops. For example, network sockets are resources. Resources may or may not correspond to a real operating system file descriptor (hence the different name).

File: core/resources.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+use downcast_rs::Downcast;
+use std;
+use std::any::Any;
+use std::collections::BTreeMap;
+use std::io::Error;
+use std::io::ErrorKind;
+
+/// Also referred to as rid.
```
Comment (ry) [inline core/resources.rs]: ResourceId is Deno's version of a file descriptor. ResourceId is also referred to as rid in the code base.

File: core/resources.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+pub type ResourceId = u32;
+
+/// These store Deno's file descriptors. These are not necessarily the operating
+/// system ones.
+type ResourceMap = BTreeMap<ResourceId, Box<dyn Resource>>;
```
Comment (ry) [inline core/resources.rs]: Why BTreeMap?

File: core/resources.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  pub fn get_mut<T: Resource>(
+    &mut self,
+    rid: ResourceId,
+  ) -> Result<&mut T, Error> {
+    let resource = self.map.get_mut(&rid).ok_or_else(bad_resource)?;
+    let resource = resource.downcast_mut::<T>().ok_or_else(bad_resource)?;
+    Ok(resource)
+  }
+
+  fn next_rid(&mut self) -> ResourceId {
```
Comment (ry) [inline core/resources.rs]: Please add `// TODO resource id allocation should probably be randomized for security.`

File: core/resources.rs (hunk 6 above)
Comment (bartlomieju) [reply core/resources.rs]: I'm not sure, I think I changed it to `BTreeMap` some time ago because of perf... Changed to `HashMap` for now

**Comment by bartlomieju** [comment]
> Looks great! Thank you for this important work!
> I defer to @piscisaureus to approve and land.

Cool, though we shouldn't land it until we figure out how to implement `read`/`write` handling

**Comment by ry** [comment]
> Cool, though we shouldn't land it until we figure out how to implement `read`/`write` handling

The way you handled it in the example is good.

**Comment by bartlomieju** [comment]
> > Cool, though we shouldn't land it until we figure out how to implement `read`/`write` handling
> 
> The way you handled it in the example is good.

Can't be translated to CLI though, PTAL at [this comment](https://github.com/denoland/deno/pull/3150#discussion_r336732647)

**Comment by ry** [comment]
> Can't be translated to CLI though, PTAL at this comment

I don't understand why it can't work in the CLI. It's the same operation.

**Pull request merged**